pkgs/tools/security/pinentry/default.nix

```diff
@@ -1,60 +1,96 @@
 { fetchurl, fetchpatch, stdenv, lib, pkgconfig, autoreconfHook
 , libgpgerror, libassuan
-, libcap ? null, libsecret ? null, ncurses ? null, gtk2 ? null, gcr ? null
-, qt4 ? null, qt5 ? null
-, enableEmacs ? false
+, ncurses, gtk2, qt
+, libcap ? null, libsecret ? null, gcr ? null
+, enabledFlavors ? [ "curses" "tty" "gtk2" "qt" "gnome3" "emacs" ]
 }:
 
-assert qt5 != null -> qt4 == null;
-assert qt4 != null -> qt5 == null;
+with stdenv.lib;
+
+assert isList enabledFlavors && enabledFlavors != [];
 
 let
   mkDerivation =
-    if qt5 != null
-      then qt5.mkDerivation
+    if (builtins.elem "qt" enabledFlavors)
+      then qt.mkDerivation
       else stdenv.mkDerivation;
+
+  mkFlag = pfxTrue: pfxFalse: cond: name:
+    "--${if cond then pfxTrue else pfxFalse}-${name}";
+  mkEnable = mkFlag "enable" "disable";
+  mkWith = mkFlag "with" "without";
+
+  mkEnablePinentry = f:
+    let
+      info = flavorInfo.${f};
+      inputs = info.buildInputs or [];
+      flag = flavorInfo.${f}.flag or null;
+      inputsSatisfied = all (f: ! isNull f) inputs;
+    in
+      assert inputsSatisfied;
+      optionalString (flag != null)
+        (mkEnable (elem f enabledFlavors && inputsSatisfied) ("pinentry-" + flag));
+
+  flavorInfo = {
+    curses = { bin = "curses"; buildInputs = [ ncurses ]; };
+    tty = { bin = "tty"; flag = "tty"; };
+    gtk2 = { bin = "gtk-2"; flag = "gtk2"; buildInputs = [ gtk2 ]; };
+    gnome3 = { bin = "gnome3"; flag = "gnome3"; buildInputs = [ gcr ]; };
+    qt = { bin = "qt"; flag = "qt"; buildInputs = [ qt.qtbase ]; };
+    emacs = { bin = "emacs"; flag = "emacs"; buildInputs = []; };
+  };
+
 in
 
 mkDerivation rec {
-  name = "pinentry-1.1.0";
+  pname = "pinentry";
+  version = "1.1.0";
 
   src = fetchurl {
-    url = "mirror://gnupg/pinentry/${name}.tar.bz2";
+    url = "mirror://gnupg/pinentry/${pname}-${version}.tar.bz2";
     sha256 = "0w35ypl960pczg5kp6km3dyr000m1hf0vpwwlh72jjkjza36c1v8";
   };
 
-  nativeBuildInputs = [ pkgconfig autoreconfHook ];
-  buildInputs =
-    [ libgpgerror libassuan libcap libsecret gtk2 gcr ncurses qt4 ]
-    ++ stdenv.lib.optional (qt5 != null) qt5.qtbase;
+  nativeBuildInputs = [ pkgconfig autoreconfHook qt.wrapQtAppsHook ];
```

Comment:
Can we add `qt.wrapQtAppsHook` only when building qt flavour? It still adds some qt deps into build closure.

Reply:
done.